### DISCUSSION TITLE
Switch CPU on the fly?

### DISCUSSION THREAD

**Discussion by udo-munk** [udo-munk/z80pack#310] [Ideas]
simh implementation Z80/8080 can switch CPU on the fly via port output, so it also is under control of a running program.
In the z80pack machines we have the hardware control port and it has unused bits left. Figure if we just can switch CPU,
or if this confuses the core emulations.

**Comment by sneakywumpus** [comment]
Is this really useful? Would this involve a CPU reset on switching? What if some code detects at runtime if it is running on a Z80 and patches itself or something like that?

**Reply by udo-munk** [reply]
After implementation I can say for sure that it is useful, also the idea to switch from the ICE is great, because now we can do this:
https://youtu.be/0Qe5j8Dla2s

**Reply by sneakywumpus** [reply]
Didn't think about that when I implemented this... there was no easy way to switch the CPU on the Pico for running 8080 only software.

**Reply by udo-munk** [reply]
Now there is, really cool stuff.

**Reply by udo-munk** [reply]
BTW, you can check the size of the executable in the build directory with:
size picosim.elf
   text    data     bss     dec     hex filename
 509772       0    6028  515800   7ded8 picosim.elf
The board has 2 MB flash and 264KB SRAM, still plenty for everything I want in there.

**Comment by udo-munk** [comment]
No reset, it continues to run on the other CPU with the next instruction. Exactely that is what EX.MAC is doing on simh, see source on cpmsim/disks/library/z80tests.dsk.
Is a nice feature, but normaly not needed. I really don't know right now how much grief this would cause, might not be worth the trouble.

**Reply by sneakywumpus** [reply]
Would this than also use port 254 to be compatible with simh? But imsaisim uses port 254 for an unimplemented memory write protect. Or only the hardware control port 160?

**Reply by udo-munk** [reply]
Right, in simh they made the wrong decision to use a port that some existing machines use for memory management. Here we would add it to the 160 hardware control.

**Reply by udo-munk** [reply]
Which is lock protected by the way, because in the past some funny game poked on it thinking there is a music card ;-)
